okapi-common/src/main/java/org/folio/okapi/common/refreshtoken/client/TokenClient.java

```diff
@@ -0,0 +1,130 @@
+package org.folio.okapi.common.refreshtoken.client;
+
+import io.netty.handler.codec.http.cookie.ClientCookieDecoder;
+import io.netty.handler.codec.http.cookie.Cookie;
+import io.vertx.core.Future;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.web.client.HttpRequest;
+import io.vertx.ext.web.client.WebClient;
+import java.util.function.Supplier;
+import org.folio.okapi.common.XOkapiHeaders;
+import org.folio.okapi.common.refreshtoken.tokencache.TokenCache;
+
+public class TokenClient {
+  private final TokenCache cache;
+  private final WebClient client;
+  private final String okapiUrl;
+  private final String tenant;
+  private final String username;
+  private final Supplier<Future<String>> getPasswordSupplier;
+
+  /**
+   * Refresh legacy tokens older than this.
+   */
+  private static final long AGE_LEGACY_TOKEN = 86400L;
+
+  /**
+   * Subtract this many seconds of age, before considering expired.
+   */
+  private static final long AGE_DIFF_TOKEN = 10L;
+
+  /**
+   * Construct client. Used normally for each incoming request.
+   * @param okapiUrl OKAPI URL for to use for getting token
+   * @param client WebClient
+   * @param cache token cache; maybe null for no cache (testing ONLY)
+   * @param tenant Okapi tenant
+   * @param username username to use for getting token
+   * @param getPasswordSupplier for providing the password
+   */
+  public TokenClient(String okapiUrl, WebClient client, TokenCache cache, String tenant,
+      String username, Supplier<Future<String>> getPasswordSupplier) {
+    this.cache = cache;
+    this.client = client;
+    this.okapiUrl = okapiUrl;
+    this.tenant = tenant;
+    this.username = username;
+    this.getPasswordSupplier = getPasswordSupplier;
+  }
+
```

Comment:
My main concern with this approach is that it removes the RT from the process. In a perfect world the RT should be the parameter here rather than the username and password. By removing RTs we're removing the security benefits that it provides. Things like expiration, revocation, and single use.

I understand why you've taken the credentials-based approach. It's pragmatic and it will "always work". If we only have the RT in the client, the RT could expire and then what happens? But this would only happen for very sleepy modules.

My thought is that solve this at a different level. Let dev ops have the responsibility of keeping the RT fresh for modules that have gone to sleep for more than the TTL of the RT. Credentials don't have live in the module at all. Just the RT. Then we get all the benefits of the RT.

How would dev ops solve this? They could generate an RT when a module gets deployed. You could have a job that goes in every day or week and refreshes them for sleepy modules. Most modules won't need this because they'll always be awake getting new RTs when they need them. If you have an RT, and you refresh it when you need to, you can always get a new one. No creds are needed.

Reply:
I'm glad code illustrates this issue. The whole idea of storing passwords with a module and using the notion of users that are tied to inventory with services points is so wrong. For example, system modules has same password for all tenants - or is this only really useful for single tenant deployments?

The RTR was for the UI. That was the idea. Not for anything else. I don't know convincing argument there was for this to be everywhere and do it all within one release. I don't see it.

I can close this PR and throw away the work I've done. Like two days getting edge-common to work this. edge-modules would also be a lot simpler if the secret was not the password but the API token!

Reply:
@steveellis @adam I assume we all agree that storing password with the module (like it happens for edge-modules and mod-pubsub) is wrong. But it also seems wrong that these modules use RTR and not API Tokens. 
But let's not try to enforce any specific approach with this PR and leave the client generic. How about we let the client of this API to specify what type of credential should be send on subsequent login requests? If they want to keep on re-logging in u/p -- fine, but if they use /login-with-expiry and want to use RT instead, let them.

Reply:
Have a look at the client now with three ways to do it.

Reply:
Adam as we've been talking about on Slack the flaw in the RT-only-based (no credential) approach for this is that it would require writable storage in each client to keep the latest RT since you get a new one each time. If you have a user name and password it can be made to work, but we're trying to make that be the case anymore.

As we've been saying too, if we want not to store usernames and passwords in the clients you could place an API token in the environment of each module. This API token would either take the palace of or allow us to get a token that takes the place of the current system user tokens.

This type of token is revokable since env can be reset. Storing secrets via mechanisms available in every deployment's env is what we're already doing. Admins will expect to be able to reset an env var via that mechanism (updates to secrets, restarts, whatever). Okapi can do this. Kubernetes too.

I have a few questions:
- Do we really care if the API token is tied to a user? What is the downside? It's what we're doing now. We don't need API tokens to have any more permissions (like module-level perms) than they do right now do we?
- Our current system user tokens _are_ tied to a certain kind of user (a "system user"). Why not just keep that structure? Less to argue about. No need to change how permissions work now. And there are quite a few special things these system users can do. Do we need to change that now?
- If we're just storing it in the env in deployments what we're going to have is tenant and maybe user-level environment variables. This number isn't huge since I think we're talking about system users. But this will need keys at least like `API_TOKEN_<tenant>`. Is that ok with us? 
- Is there any value to using this api token as a token that only gets access tokens? Skott's review suggested that API tokens not be presentable as access tokens and that's what we'd be doing here, unless we implemented a "refresh" mechanism where api tokens are used only to get new short-lived ATs. Being able to instantly revoke via the env though may be _more_ secure.